<a href="https://colab.research.google.com/github/neilsambhu/carla-simulator2/blob/main/carla_simulator2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Sandbox for CARLA simulator following https://github.com/MichaelBosello/carla-colab. Original *.ipynb file: https://colab.research.google.com/github/MichaelBosello/carla-colab/blob/master/carla-simulator.ipynb. 

In case of poor functionality: Runtime > Restart runtime (Ctrl+M.)

In case of very poor functionality: Runtime > Factory reset runtime

2/17/2022 8:49 PM: instead of (1) "!" to execute shell commands, use (2) the Python "subprocess" module to ensure compatibility between iPython and Python (complete as of 2/17/2022 10:16 PM).

2/17/2022 10:23 PM: I made the Call_SubprocessPopen(cmd) function for easily calling shell commands. 

2/17/2022 11:02 PM: The subprocess.run will fail for command-line arguments separated by spaces. The subprocess.run requires a list.

In [1]:
import subprocess, sys
from sys import platform
def Call_SubprocessPopen(cmd):
  p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
  # p = subprocess.run(cmd, capture_output=True, text=True, check=True)
  for line in p.stdout.readlines():
    sys.stdout.write(line.decode("utf-8"))
    # sys.stdout.write((r"utf-8")line)
    pass
  # print(p.stdout)
if platform == "linux" or platform == "linux2":
  # for multi-line output: use "pip --help" instead of "pwd" 
  cmd = "pwd"
  # cmd = "pip --help"
  # cmd = "nvidia-smi"
  Call_SubprocessPopen(cmd)
elif platform == "win32":
  raise Exception("2/17/2022 8:59 PM: This code is designed to run in a Linux-based environment.")

/content


In [2]:
# <10 seconds
Call_SubprocessPopen("pip install git+https://github.com/demotomohiro/remocolab.git")

  Cloning https://github.com/demotomohiro/remocolab.git to /tmp/pip-req-build-zbp59ixz
  Running command git clone -q https://github.com/demotomohiro/remocolab.git /tmp/pip-req-build-zbp59ixz
  Created wheel for remocolab.py: filename=remocolab.py-0.1-py3-none-any.whl size=7553 sha256=c5211d4f40bf8f13b809fb6a13a94c86333b773de3b688abb502e830977dbe97
  Stored in directory: /tmp/pip-ephem-wheel-cache-f3_bwb4p/wheels/80/37/44/386e71ab0cd5aef5cefdb1926b710a42fdb4fe2d07c3b1b08f
  Created wheel for pyngrok: filename=pyngrok-5.1.0-py3-none-any.whl size=19007 sha256=f9adeec231fbf4eefc54eb10382f897c28f736ed3fea5f28fdfdcf84e0a68c6f
  Stored in directory: /root/.cache/pip/wheels/bf/e6/af/ccf6598ecefecd44104069371795cb9b3afbcd16987f6ccfb3
Successfully built remocolab.py pyngrok


In [3]:
# <5.5 minutes
import remocolab
remocolab.setupVNC()

Warning! GPU of your assigned virtual machine is Tesla K80.
You might get better GPU by reseting the runtime.
Do you want to continue? [y/n] y
---
As ngrok doesn't work on colab for long time, default tunnel method has been changed to Argo tunnel.
Please read this for more details:
https://github.com/demotomohiro/remocolab/blob/master/README.md


Label(value='')

FloatProgress(value=0.0, layout=Layout(border='1px solid #118800'), max=1.0)

Install openssh-server


Label(value='')

FloatProgress(value=0.0, layout=Layout(border='1px solid #118800'), max=1.0)

Install xfce4
Install xfce4-terminal
ED25519 key fingerprint of host:
256 SHA256:Fupwfh7kRisd6OLBJVK5UAoSBejW7+FQUdtclZE3S5I root@44bf28f802a0 (ED25519)
+--[ED25519 256]--+
|*+. . ..   ..o=  |
|o. o o  + .  E + |
|. + o ...o    + o|
| o + o o .     . |
|. . * = S        |
|   + @ * o       |
|    B * B        |
|   . = = .       |
|    .   .        |
+----[SHA256]-----+

✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
root password: YSWo61RYq-jFiCyugi8ppGDnnTQmEkY-Of-bz_qYimw
colab password: N9L24iU6pmaD0AxGgkfeL7J1xezCe4TfPzU20krnyWQ
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
---
Execute following command on your local machine and login before running TurboVNC viewer:
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -oProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@concentrations-year-ceremony-ctrl.trycloudflare.com
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
VNC password: msDX77sx
VN

2/20/2022 9:01 PM: Problem:
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -oProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@vat-bracket-mean-hobbies.trycloudflare.com
/bin/bash: line 0: exec: cloudflared: not found
kex_exchange_identification: Connection closed by remote host

[install cloudflared (https://pkg.cloudflare.com/#ubuntu-title) in Windows Subsystem for Linux (WSL) Ubuntu 20.04 LTS]

2/20/2022 9:08 PM: Problem
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -o ProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@vat-bracket-mean-hobbies.trycloudflare.com
/bin/bash: line 0: exec: cloudflared: not found
kex_exchange_identification: Connection closed by remote host

[WSL: sudo apt install cloudflared]

In [4]:
# <4 minutes

# def IOPub_DataRate(iopub_data_rate_limit=1000000.0, rate_limit_window=3.0):
#   import os
#   # os.environ['NotebookApp.iopub_data_rate_limit']=str(iopub_data_rate_limit)
#   Call_SubprocessPopen("--NotebookApp.iopub_data_rate_limit {}".format(iopub_data_rate_limit))
#   os.environ['NotebookApp.rate_limit_window']=str(rate_limit_window)
# # increase IOPub data rate; set the config variable `--NotebookApp.iopub_data_rate_limit`.
# # Current values:
# # NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
# # NotebookApp.rate_limit_window=3.0 (secs)
# IOPub_DataRate(1e7)

# Download CARLA
# Call_SubprocessPopen("sudo -u colab wget -O /home/colab/CARLA.tar.gz https://carla-releases.s3.eu-west-3.amazonaws.com/Linux/CARLA_0.9.13.tar.gz")
# 2/22/2022 3:23 PM: try CARLA 0.9.10 instead of 0.9.13
Call_SubprocessPopen("sudo -u colab wget -O /home/colab/CARLA.tar.gz https://carla-releases.s3.eu-west-3.amazonaws.com/Linux/CARLA_0.9.10.tar.gz")

# reset IOPub data rate
# IOPub_DataRate()

Streaming output truncated to the last 5000 lines.
1318750K .......... .......... .......... .......... .......... 34% 11.5M 89s
1318800K .......... .......... .......... .......... .......... 34% 21.6M 89s
1318850K .......... .......... .......... .......... .......... 34%  135M 89s
1318900K .......... .......... .......... .......... .......... 34%  154M 89s
1318950K .......... .......... .......... .......... .......... 34% 60.8M 89s
1319000K .......... .......... .......... .......... .......... 34%  155M 89s
1319050K .......... .......... .......... .......... .......... 34% 9.86M 89s
1319100K .......... .......... .......... .......... .......... 34% 56.4M 89s
1319150K .......... .......... .......... .......... .......... 34% 27.8M 89s
1319200K .......... .......... .......... .......... .......... 34%  143M 89s
1319250K .......... .......... .......... .......... .......... 34% 26.3M 89s
1319300K .......... .......... .......... .......... .......... 34%  144M 89s
1319350K ....

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [5]:
# <3.5 minutes
Call_SubprocessPopen("sudo -u colab mkdir /home/colab/carla")
Call_SubprocessPopen("sudo -u colab tar -xf /home/colab/CARLA.tar.gz -C /home/colab/carla")

2/20/2022 10:41 PM: 

Disk is almost full

Your disk is almost full. Completely filling the disk will result in errors. Currently, 63.39 GB / 78.19 GB is being used. You can delete files in order to free up disk space. If you are interested in Colab VMs with larger disks, you may want to check out Colab Pro.

Setup PythonAPI

In [6]:
# <2 seconds
Call_SubprocessPopen("sudo -u colab rm -rf /home/colab/carla/PythonAPI/carla/dist/*.egg")
Call_SubprocessPopen("sudo -u colab wget https://github.com/MichaelBosello/carla-colab/raw/master/carla-0.9.10-py3.6-linux-x86_64.egg -P /home/colab/carla/PythonAPI/carla/dist/")

--2022-02-25 03:08:17--  https://github.com/MichaelBosello/carla-colab/raw/master/carla-0.9.10-py3.6-linux-x86_64.egg
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/MichaelBosello/carla-colab/master/carla-0.9.10-py3.6-linux-x86_64.egg [following]
--2022-02-25 03:08:18--  https://raw.githubusercontent.com/MichaelBosello/carla-colab/master/carla-0.9.10-py3.6-linux-x86_64.egg
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 23488661 (22M) [application/octet-stream]
Saving to: ‘/home/colab/carla/PythonAPI/carla/dist/carla-0.9.10-py3.6-linux-x86_64.egg’

     0K .......... .......... .......... .....

Run CARLA

In [7]:
# <6 seconds
Call_SubprocessPopen("apt-get install libomp5")

Reading package lists...
Building dependency tree...
Reading state information...
The following packages were automatically installed and are no longer required:
  cuda-command-line-tools-10-0 cuda-command-line-tools-10-1
  cuda-command-line-tools-11-0 cuda-compiler-10-0 cuda-compiler-10-1
  cuda-compiler-11-0 cuda-cuobjdump-10-0 cuda-cuobjdump-10-1
  cuda-cuobjdump-11-0 cuda-cupti-10-0 cuda-cupti-10-1 cuda-cupti-11-0
  cuda-cupti-dev-11-0 cuda-documentation-10-0 cuda-documentation-10-1
  cuda-documentation-11-0 cuda-documentation-11-1 cuda-gdb-10-0 cuda-gdb-10-1
  cuda-gdb-11-0 cuda-gpu-library-advisor-10-0 cuda-gpu-library-advisor-10-1
  cuda-libraries-10-0 cuda-libraries-10-1 cuda-libraries-11-0
  cuda-memcheck-10-0 cuda-memcheck-10-1 cuda-memcheck-11-0 cuda-nsight-10-0
  cuda-nsight-10-1 cuda-nsight-11-0 cuda-nsight-11-1 cuda-nsight-compute-10-0
  cuda-nsight-compute-10-1 cuda-nsight-compute-11-0 cuda-nsight-compute-11-1
  cuda-nsight-systems-10-1 cuda-nsight-systems-11-0 cuda-nsig

In [8]:
# Call_SubprocessPopen("sudo -u colab DISPLAY=:1 vglrun /home/colab/carla/CarlaUE4.sh")

2/20/2022 10:50 PM:

TurboVNC: 

The global shader cache file '/home/colab/carla/Engine/GlobalShaderCache-GLSL_150_ES31.bin' is missing.

Your application is built to load COOKED content. No COOKED content was found; This usually means you did not cook content for this build. It also may indicate missing cooked data for a shader platform(e.g., OpenGL under Windows); Make sure your platform's packaging settings include this Targeted RHI.

Alternatively build and run the UNCOOKED version instead.

2/22/2022 3:41 PM: This error disappears for CARLA 0.9.10 instead of 0.9.13.

Client on Colab

In [9]:
# <13 minutes
Call_SubprocessPopen("sudo pip3 install pygame")
# Install CARLA
Call_SubprocessPopen("sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 1AF1527DE64CB8D9")
Call_SubprocessPopen('sudo add-apt-repository "deb [arch=amd64] http://dist.carla.org/carla $(lsb_release -sc) main"')
Call_SubprocessPopen("sudo apt-get update")
Call_SubprocessPopen("sudo apt-get install carla-simulator=0.9.10-1")

# Call_SubprocessPopen("sudo -u colab DISPLAY=:1 vglrun /home/colab/carla/CarlaUE4.sh && python3 /home/colab/carla/PythonAPI/examples/manual_control.py")


Executing: /tmp/apt-key-gpghome.yMe0cxSNSI/gpg.1.sh --keyserver keyserver.ubuntu.com --recv-keys 1AF1527DE64CB8D9
gpg: key 1AF1527DE64CB8D9: public key "Carla Simulator Team <carla.simulator@gmail.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1
Hit:1 http://security.ubuntu.com/ubuntu bionic-security InRelease
Hit:2 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease
Ign:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
Ign:4 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Get:5 http://dist.carla.org/carla bionic InRelease [2,338 B]
Hit:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release
Hit:7 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Hit:8 http://archive.ubuntu.com/ubuntu bionic InRelease
Hit:9 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease
Get

In [10]:
Call_SubprocessPopen("cat /home/colab/carla/CarlaUE4.sh")

#!/bin/sh
UE4_TRUE_SCRIPT_NAME=$(echo \"$0\" | xargs readlink -f)
UE4_PROJECT_ROOT=$(dirname "$UE4_TRUE_SCRIPT_NAME")
chmod +x "$UE4_PROJECT_ROOT/CarlaUE4/Binaries/Linux/CarlaUE4-Linux-Shipping"
"$UE4_PROJECT_ROOT/CarlaUE4/Binaries/Linux/CarlaUE4-Linux-Shipping" CarlaUE4 $@ 


In [11]:
Call_SubprocessPopen("cat /home/colab/carla/PythonAPI/examples/manual_control.py")

#!/usr/bin/env python

# Copyright (c) 2019 Computer Vision Center (CVC) at the Universitat Autonoma de
# Barcelona (UAB).
#
# This work is licensed under the terms of the MIT license.
# For a copy, see <https://opensource.org/licenses/MIT>.

# Allows controlling a vehicle with a keyboard. For a simpler and more
# documented example, please take a look at tutorial.py.

"""
Welcome to CARLA manual control.

Use ARROWS or WASD keys for control.

    W            : throttle
    S            : brake
    A/D          : steer left/right
    Q            : toggle reverse
    Space        : hand-brake
    P            : toggle autopilot
    M            : toggle manual transmission
    ,/.          : gear up/down
    CTRL + W     : toggle constant velocity mode at 60 km/h

    L            : toggle next light type
    SHIFT + L    : toggle high beam
    Z/X          : toggle right/left blinker
    I            : toggle interior light

    TAB          : change sensor position
    ` or N       

2/22/2022 9:24 PM: Error in manual_control.py > game_loop > display

display = pygame.display.set_mode(
            (args.width, args.height),
            pygame.HWSURFACE | pygame.DOUBLEBUF)

In [12]:
Call_SubprocessPopen("python3 /home/colab/carla/PythonAPI/examples/manual_control.py")

INFO: listening to server 127.0.0.1:2000
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default
pygame 2.1.2 (SDL 2.0.16, Python 3.7.12)
Hello from the pygame community. https://www.pygame.org/contribute.html

Welcome to CARLA manual control.

Use ARROWS or WASD keys for control.

    W            : throttle
    S            : brake
    A/D          : steer left/right
    Q      